selfservice/hook/web_hook.go

```diff
@@ -274,29 +274,36 @@ func (e *WebHook) execute(ctx context.Context, data *templateContext) error {
 			attribute.String("webhook.identity.nid", data.Identity.NID.String()),
 		)
 	}
+
 	var (
-		httpClient                       = e.deps.HTTPClient(ctx)
-		async                            = gjson.GetBytes(e.conf, "response.ignore").Bool()
-		parseResponse                    = gjson.GetBytes(e.conf, "can_interrupt").Bool()
-		tracer                           = trace.SpanFromContext(ctx).TracerProvider().Tracer("kratos-webhooks")
-		cancel        context.CancelFunc = func() {}
-		spanOpts                         = []trace.SpanStartOption{trace.WithAttributes(attrs...)}
-		errChan                          = make(chan error, 1)
+		httpClient     = e.deps.HTTPClient(ctx)
+		ignoreResponse = gjson.GetBytes(e.conf, "response.ignore").Bool()
+		canInterrupt   = gjson.GetBytes(e.conf, "can_interrupt").Bool()
+		tracer         = trace.SpanFromContext(ctx).TracerProvider().Tracer("kratos-webhooks")
+		spanOpts       = []trace.SpanStartOption{trace.WithAttributes(attrs...)}
+		errChan        = make(chan error, 1)
 	)
-	if async {
-		// dissociate the context from the one passed into this function
-		ctx, cancel = context.WithTimeout(context.Background(), 5*time.Minute)
-		spanOpts = append(spanOpts, trace.WithNewRoot())
-	}
+
 	ctx, span := tracer.Start(ctx, "Webhook", spanOpts...)
 	e.deps.Logger().WithRequest(req.Request).Info("Dispatching webhook")
-	t0 := time.Now()
+
+	req = req.WithContext(ctx)
+	if ignoreResponse {
+		// This is one of the few places where spawning a context.Background() is ok. We need to do this
+		// because the function runs asynchronously and we don't want to cancel the request if the
+		// incoming request context is cancelled.
+		//
+		// The webhook will still cancel after 30 seconds as that is the configured timeout for the HTTP client.
+		req = req.WithContext(context.Background())
```

Comment:
Instead of overshadowing the context, we now use `context.Background()` only where it is needed - in the request context. An alternative would be to not include a context if the request is async:

```req = req.WithContext(ctx)
 	if !ignoreResponse {
 	 	req = req.WithContext(ctx)
 	}
```